experiment/nursery/main.go

```diff
@@ -0,0 +1,136 @@
+// supporting arbitrarily many workspaces stored within the same
+// top level directory.
+//
+// the first path segment in each {PUT,GET} request is mapped to an individual
+// workspace cache, the remaining segments should follow [2].
+//
+// nursery assumes you are using SHA256
+//
+// [1] https://docs.bazel.build/versions/master/remote-caching.html
+// [2] https://docs.bazel.build/versions/master/remote-caching.html#http-caching-protocol
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"k8s.io/test-infra/experiment/nursery/diskcache"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var dir = flag.String("dir", "", "location to store cache entries on disk")
+var host = flag.String("host", "", "host address to listen on")
+var port = flag.Int("port", 8080, "port to listen on")
+
+func init() {
+	log.SetFormatter(&log.TextFormatter{})
+	log.SetOutput(os.Stdout)
+}
+
+func main() {
+	// TODO(bentheelder): bound cache size / convert to LRU
+	// TODO(bentheelder): improve logging
+	flag.Parse()
+	if *dir == "" {
+		log.Fatal("--dir must be set!")
+	}
+
+	cache := diskcache.NewCache(*dir)
+	http.Handle("/", cacheHandler(cache))
+
+	addr := fmt.Sprintf("%s:%d", *host, *port)
+	log.Infof("Listening on: %s", addr)
+	log.WithError(http.ListenAndServe(addr, nil)).Fatal("ListenAndServe returned.")
+}
+
+// file not found error, used below
+var errNotFound = errors.New("entry not found")
+
+func cacheHandler(cache *diskcache.Cache) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// parse and validate path
+		// the last segment should be a hash, and the
+		// the second to last segment should be "ac" or "cas"
+		parts := strings.Split(r.URL.Path, "/")
+		if len(parts) < 3 {
+			log.Warn("received an invalid request at path: %v", r.URL.Path)
+			http.Error(w, "invalid location", http.StatusBadRequest)
+			return
+		}
+		hash := parts[len(parts)-1]
+		acOrCAS := parts[len(parts)-2]
+		if acOrCAS != "ac" && acOrCAS != "cas" {
+			log.Warn("received an invalid request at path: %v", r.URL.Path)
+			http.Error(w, "invalid location", http.StatusBadRequest)
+			return
+		}
+
+		// actually handle request depending on method
+		switch m := r.Method; m {
+		// handle retreival
+		case http.MethodGet:
+			err := cache.Get(r.URL.Path, func(exists bool, contents io.ReadSeeker) error {
+				if !exists {
+					return errNotFound
+				}
+				http.ServeContent(w, r, "", time.Time{}, contents)
+				return nil
+			})
+			if err != nil {
+				// file not present
+				if err == errNotFound {
+					http.Error(w, err.Error(), http.StatusNotFound)
+					return
+				}
+				// unknown error
+				log.WithError(err).Error("error getting key")
+				http.Error(w, err.Error(), http.StatusNotFound)
+				return
+			}
+
+		// handle upload
+		case http.MethodPut:
+			// only hash CAS, not action cache
+			// the action cache is hash -> metadata
+			// the CAS is well, a CAS, which we can hash...
```

Comment:
I don't understand what the action cache (`ac`) variant is supposed to do. Where is the `hash -> metadata` mapping mentioned in the comment and how is it used?

Reply:
See the protocol linked at the top of the file and the comment below: https://docs.bazel.build/versions/master/remote-caching.html#http-caching-protocol

Reply:
Bazel does `PUT $HTTP_CACHE_URL/ac/$HASH` (bytes of PUT are action metadata) and `GET` the same way
It also does `PUT $HTTP_CACHE_URL/cas/$HASH` (bytes of PUT are content-address object storage) and `GET` the same way.